# App profitability in Apple App Store and Google Play Store

## About
For this project we assume that we're working for a company making mobile apps for iOS and Android devices, placing them in AppStore and Play Store storefronts respectively. Apps are offered for free for an english speaking audience and our main revenue is through advertisements placed in the app. App profitability is closely tied with the number of users the app has.
Goal of this excersise is to determine which apps tend to be the most profitable and are likely to succeed in the future.

## Data
This project uses two datasets:
 - Dataset of 10k apps from Google Play Store https://www.kaggle.com/datasets/lava18/google-play-store-apps
 - Dataset of 7.2k apps from Apple AppStore https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps

In [1]:
from csv import reader

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    #function that will print rows from datastet starting at [start] and ending at [end] indexes
    #function will also print number of rows and columns if optional parameter is True
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
def read_from_csv(path):
    #function that will open a csv file and process it into two lists
    #first list being the header row of the dataset, the second being list of lists containing the whole dataset
    opened_file = open(path, encoding='utf8') #Added encoding for utf8 since it gives UnicodeDecodeError otherwise.
    read_file = reader(opened_file)
    apps_data = list(read_file)
    return apps_data[0], apps_data[1:] #dataset[1:] ignores the header row and starts at the first data entry

In [4]:
#importing apple app store data
apple_apps_header, apple_apps_data = read_from_csv(r'C:\Users\ricar\Documents\Datasets\Project_1\AppleStore.csv')
#importing google play store data
google_apps_header, google_apps_data = read_from_csv(r'C:\Users\ricar\Documents\Datasets\Project_1\googleplaystore.csv')

In [5]:
#exploring the apple app store dataset
print(apple_apps_header, "\n")
explore_data(apple_apps_data, 0, 3, True) 

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


Number of rows: 7197
Number of columns: 17


We can see that the Apple app store dataset has 7197 rows, and 17 columns. Upon glancing at the data we can say that 'price', 'prime_genre' and 'rating_count_tot' colums might provide us with useful information for further analysis.

In [6]:
#exploring the google play store dataset
print(google_apps_header, "\n")
explore_data(google_apps_data, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


Google Play store dataset is larger in terms of rows, 108041 rows, and while it has less columns than the Apple datasource we can see, that it has more useful columns here the :'Category', 'Rating', 'Reviews', 'Installs', 'Type', 'Price' and 'Genre' columns might provide us with useful information.

## Data Cleanup and preparation
In the about section we made two assumptions:
 - Apps we make are free
 - Apps we make are for an english-speaking audience

Provided these assumptions we need to cleanup any data that does not mach our criteria, as well as any incorrect data.
### Pt.1 Cleaning up incorrect data:
Looking into the source of the Google Play store dataset in kaggle.com we can find this in the discussion section for the dataset. https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015

In [7]:
print(google_apps_header, "\n")
explore_data(google_apps_data, 10471, 10474, False)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']




Here we can see that 'Life Made WI-Fi Touchscreen Photo Frame' app is missing it's 'Category' entry, thus all other entries are shifted to the left, 'Rating' is in the place of 'Category', 'Reviews' is in place of 'Rating', and so on.

Let's check if any other value in the dataset is missing one or more column entries. Looking at the header, we can determine the length that each row should be, in this case each row in the Apple datasource should be 17 entries long, while the Google datasource 13 entries long.

To analyze both datasets let's write a function that checks the length of each row.

In [8]:
def check_line_entry(dataset, line_length):
    #function takes two parameters the dataset, and expected row length
    i = 0
    list_of_indexes = []
    #making a list of row indexes, that do not match the expected row length
    for entry in dataset:
        if len(entry) != line_length:
            list_of_indexes.append(i)
        i += 1
    return list_of_indexes

Using the function above let's check the Google Play store dataset for other rows with missing data.

In [9]:
google_bad_index = check_line_entry(google_apps_data, len(google_apps_header))
print(google_bad_index)

[10472]


It appears that only one data entry was missing it's data.  Having this information we can safely remove this entry from our dataset. For that let's write a function, that we can later re-use with the Apple dataset as well.

In [10]:
def remove_entries(dataset, list_of_index):
    #function takes the dataset and list of indexes, for which rows need to be removed
    removed = 0
    for ind in list_of_index:
        del dataset[ind - removed] #Counting how many items were removed and adjusting the index accordingly.
        removed += 1 

In [11]:
remove_entries(google_apps_data, google_bad_index)

Checking if the removal of the incorrect values was successful. It appears so, row containing data for 'Life Made WI-Fi Touchscreen Photo Frame' app no longer exists.

In [12]:
explore_data(google_apps_data, 10471, 10474, False)

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


['Sat-Fi Voice', 'COMMUNICATION', '3.4', '37', '14M', '1,000+', 'Free', '0', 'Everyone', 'Communication', 'November 21, 2014', '2.2.1.5', '2.2 and up']




Let's try the same approach for the Apple app store dataset.

In [13]:
apple_bad_index = check_line_entry(apple_apps_data, len(apple_apps_header))
print(apple_bad_index)

[]


Appears that in the Apple AppStore dataset no data entries are missing values.
### Pt. 2 Cleaning up duplicate data
Having made sure no incorrect data remains in both datasources, the next step in preparing the data is to clean up any duplicate values. 

We need to check for duplicate app names in both datasources. The question then becomes if we find two apps with the same name which one we should keep. For this we will use the 'rating_count_tot' column in the Apple dataset and the 'Reviews'. This will give us a unique identifier for the app and we will keep the entry that has the most ratings/reviews. For that let's write a fuction that can be applied for both datasets.

In [14]:
def check_duplicate_names(dataset, name_column_id, review_column_id):
    #function that takes the datasource, name column id, and review count column id
    unique_name = {}
    #creating a dictionary, that will use app name as key, and the greateast amount of reviews as value
    for entry in dataset:
        if entry[name_column_id] in unique_name:
            if unique_name[entry[name_column_id]] < float(entry[review_column_id]):
                unique_name[entry[name_column_id]] = float(entry[review_column_id])
        else:
            unique_name[entry[name_column_id]] = float(entry[review_column_id])
    return unique_name #id_of_duplicates

In [15]:
unique_google = check_duplicate_names(google_apps_data, 0, 3)

In [16]:
unique_apple = check_duplicate_names(apple_apps_data, 2, 6)

In [17]:
def cleanup_duplicates(unique_dictionary, dataset, name_col_id, review_col_id):
    clean = []
    already_in_dataset = []
    for entry in dataset:
        name = entry[name_col_id]
        reviews = entry[review_col_id]
        if name not in already_in_dataset:
            if (name not in clean) and (float(reviews) == unique_dictionary[name]):
                clean.append(entry)
                already_in_dataset.append(name)
                
    return clean

In [18]:
unique_google_data = cleanup_duplicates(unique_google, google_apps_data, 0, 3)

In [19]:
unique_apple_data = cleanup_duplicates(unique_apple, apple_apps_data, 2, 6)

In [20]:
def is_name_english(name):
    non_english_letters = 0
    for letter in name:
        if ord(letter) > 127:
            non_english_letters += 1
        
    if non_english_letters > 3:
        return False
    else:
        return True

In [21]:
is_name_english('Instagram')

True

In [22]:
is_name_english('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [23]:
is_name_english('Docs To Go™ Free Office Suite')

True

In [24]:
is_name_english('Instachat 😜')

True

In [25]:
def cleanup_non_english(dataset, name_col_id):
    english_apps = []
    for entry in dataset:
        name = entry[name_col_id]
        if is_name_english(name):
            english_apps.append(entry)
    
    return english_apps

In [26]:
english_google_data = cleanup_non_english(unique_google_data, 0)

In [27]:
english_apple_data = cleanup_non_english(unique_apple_data, 2)

In [28]:
explore_data(english_google_data, 0, 2, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9614
Number of columns: 13


In [29]:
explore_data(english_apple_data, 0, 2, True)

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


Number of rows: 6181
Number of columns: 17


In [30]:
def convert_price_to_float(price_str):
    if price_str[:1] == '$':
        return float(price_str[1:])
    else:
        return float(price_str)

In [31]:
def isolate_free(dataset, price_col_id):
    free_apps = []
    for entry in dataset:
        price = convert_price_to_float(entry[price_col_id])
        if price == 0:
            free_apps.append(entry)
            
    return  free_apps

In [32]:
free_google_data = isolate_free(english_google_data, 7)
explore_data(free_google_data, 0, 2, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 8864
Number of columns: 13


In [33]:
free_apple_data = isolate_free(english_apple_data, 5)
explore_data(free_apple_data, 0, 2, True)

['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


Number of rows: 3220
Number of columns: 17


In [34]:
def data_frequency(dataset, frequency_column_id):
    frequency_table = {}
    entry_count = len(dataset)
    for entry in dataset:
        frequency_value = entry[frequency_column_id]
        if frequency_value in frequency_table:
            frequency_table[frequency_value] += 1
        else:
            frequency_table[frequency_value] = 1
    
    for key in frequency_table:
        frequency_table[key] = float("{:.2f}".format((frequency_table[key] * 100) / entry_count))
    return frequency_table

In [35]:
google_genre_frequency = data_frequency(free_google_data, 9)

In [36]:
apple_genre_frequency = data_frequency(free_apple_data, 12)
print(apple_genre_frequency)

{'Productivity': 1.74, 'Weather': 0.87, 'Shopping': 2.61, 'Reference': 0.56, 'Finance': 1.12, 'Music': 2.05, 'Utilities': 2.52, 'Travel': 1.24, 'Social Networking': 3.29, 'Sports': 2.14, 'Health & Fitness': 2.02, 'Games': 58.14, 'Food & Drink': 0.81, 'News': 1.34, 'Book': 0.43, 'Photo & Video': 4.97, 'Entertainment': 7.89, 'Business': 0.53, 'Lifestyle': 1.58, 'Education': 3.66, 'Navigation': 0.19, 'Medical': 0.19, 'Catalogs': 0.12}


In [37]:
def display_table(dataset, index):
    table = data_frequency(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0], '%')

In [38]:
display_table(free_google_data, 9)

Tools : 8.45 %
Entertainment : 6.07 %
Education : 5.35 %
Business : 4.59 %
Productivity : 3.89 %
Lifestyle : 3.89 %
Finance : 3.7 %
Medical : 3.53 %
Sports : 3.46 %
Personalization : 3.32 %
Communication : 3.24 %
Action : 3.1 %
Health & Fitness : 3.08 %
Photography : 2.94 %
News & Magazines : 2.8 %
Social : 2.66 %
Travel & Local : 2.32 %
Shopping : 2.25 %
Books & Reference : 2.14 %
Simulation : 2.04 %
Dating : 1.86 %
Arcade : 1.85 %
Video Players & Editors : 1.77 %
Casual : 1.76 %
Maps & Navigation : 1.4 %
Food & Drink : 1.24 %
Puzzle : 1.13 %
Racing : 0.99 %
Role Playing : 0.94 %
Libraries & Demo : 0.94 %
Auto & Vehicles : 0.93 %
Strategy : 0.91 %
House & Home : 0.82 %
Weather : 0.8 %
Events : 0.71 %
Adventure : 0.68 %
Comics : 0.61 %
Beauty : 0.6 %
Art & Design : 0.6 %
Parenting : 0.5 %
Card : 0.45 %
Casino : 0.43 %
Trivia : 0.42 %
Educational;Education : 0.39 %
Board : 0.38 %
Educational : 0.37 %
Education;Education : 0.34 %
Word : 0.26 %
Casual;Pretend Play : 0.24 %
Music : 0.2 %
R

In [39]:
display_table(free_google_data, 1)

FAMILY : 18.91 %
GAME : 9.72 %
TOOLS : 8.46 %
BUSINESS : 4.59 %
LIFESTYLE : 3.9 %
PRODUCTIVITY : 3.89 %
FINANCE : 3.7 %
MEDICAL : 3.53 %
SPORTS : 3.4 %
PERSONALIZATION : 3.32 %
COMMUNICATION : 3.24 %
HEALTH_AND_FITNESS : 3.08 %
PHOTOGRAPHY : 2.94 %
NEWS_AND_MAGAZINES : 2.8 %
SOCIAL : 2.66 %
TRAVEL_AND_LOCAL : 2.34 %
SHOPPING : 2.25 %
BOOKS_AND_REFERENCE : 2.14 %
DATING : 1.86 %
VIDEO_PLAYERS : 1.79 %
MAPS_AND_NAVIGATION : 1.4 %
FOOD_AND_DRINK : 1.24 %
EDUCATION : 1.16 %
ENTERTAINMENT : 0.96 %
LIBRARIES_AND_DEMO : 0.94 %
AUTO_AND_VEHICLES : 0.93 %
HOUSE_AND_HOME : 0.82 %
WEATHER : 0.8 %
EVENTS : 0.71 %
PARENTING : 0.65 %
ART_AND_DESIGN : 0.64 %
COMICS : 0.62 %
BEAUTY : 0.6 %


In [40]:
display_table(free_apple_data, 12)

Games : 58.14 %
Entertainment : 7.89 %
Photo & Video : 4.97 %
Education : 3.66 %
Social Networking : 3.29 %
Shopping : 2.61 %
Utilities : 2.52 %
Sports : 2.14 %
Music : 2.05 %
Health & Fitness : 2.02 %
Productivity : 1.74 %
Lifestyle : 1.58 %
News : 1.34 %
Travel : 1.24 %
Finance : 1.12 %
Weather : 0.87 %
Food & Drink : 0.81 %
Reference : 0.56 %
Business : 0.53 %
Book : 0.43 %
Navigation : 0.19 %
Medical : 0.19 %
Catalogs : 0.12 %


## App Data insights
### Apple App store
Looking at the free, english apps from the apple app store the vast majority (58.14%) of the apps are games. Close second, but only a fraction of the first are Entertainment (7,89%), and almost half that in the third place we have Photo & video apps(4,97%).
### Google play store
Looking at the Google Play store data majority of the apps fell into the three categories Family (18.91%), Game (9.72%) and Tools (8.46%), and the most dominating genres were Tools (8.45%), Entertainment (6.07%) and Education (5.35%)
### Insights
Looking at both datasets we can arrive at the conclusion, that Games, that are family friendly, or educational would fall into the most popular app category, and would provide the most success.

In [41]:
def convert_users_to_int(users):
    if users[-1] == '+':
        return(int(users[:-1].replace(',' , '')))
    else: 
        return(int(users.replace(',' , '')))

In [42]:
def avg_users_per_genre(dataset, genre_id, user_id):
    users_per_genre = {}
    frequency = {}
    for entry in dataset:
        key = entry[genre_id]
        if key in users_per_genre:
            users_per_genre[key] += convert_users_to_int(entry[user_id])
            frequency[key] += 1
        else:
            users_per_genre[key] = convert_users_to_int(entry[user_id])
            frequency[key] = 1
    for key in users_per_genre:
        users_per_genre[key] = float("{:.2f}".format(users_per_genre[key] / frequency[key]))
        
    return users_per_genre
           

In [43]:
avg_users_per_genre(free_google_data, 1, 5)

{'ART_AND_DESIGN': 1986335.09,
 'AUTO_AND_VEHICLES': 647317.82,
 'BEAUTY': 513151.89,
 'BOOKS_AND_REFERENCE': 8767811.89,
 'BUSINESS': 1712290.15,
 'COMICS': 817657.27,
 'COMMUNICATION': 38456119.17,
 'DATING': 854028.83,
 'EDUCATION': 1833495.15,
 'ENTERTAINMENT': 11640705.88,
 'EVENTS': 253542.22,
 'FINANCE': 1387692.48,
 'FOOD_AND_DRINK': 1924897.74,
 'HEALTH_AND_FITNESS': 4188821.99,
 'HOUSE_AND_HOME': 1331540.56,
 'LIBRARIES_AND_DEMO': 638503.73,
 'LIFESTYLE': 1437816.27,
 'GAME': 15588015.6,
 'FAMILY': 3695641.82,
 'MEDICAL': 120550.62,
 'SOCIAL': 23253652.13,
 'SHOPPING': 7036877.31,
 'PHOTOGRAPHY': 17840110.4,
 'SPORTS': 3638640.14,
 'TRAVEL_AND_LOCAL': 13984077.71,
 'TOOLS': 10801391.3,
 'PERSONALIZATION': 5201482.61,
 'PRODUCTIVITY': 16787331.34,
 'PARENTING': 542603.62,
 'WEATHER': 5074486.2,
 'VIDEO_PLAYERS': 24727872.45,
 'NEWS_AND_MAGAZINES': 9549178.47,
 'MAPS_AND_NAVIGATION': 4056941.77}

In [44]:
avg_users_per_genre(free_google_data, 9, 5)

{'Art & Design': 2122850.94,
 'Art & Design;Creativity': 285000.0,
 'Auto & Vehicles': 647317.82,
 'Beauty': 513151.89,
 'Books & Reference': 8767811.89,
 'Business': 1712290.15,
 'Comics': 831873.15,
 'Comics;Creativity': 50000.0,
 'Communication': 38456119.17,
 'Dating': 854028.83,
 'Education': 550185.44,
 'Education;Creativity': 2875000.0,
 'Education;Education': 4759517.0,
 'Education;Pretend Play': 1800000.0,
 'Education;Brain Games': 5333333.33,
 'Entertainment': 5602792.78,
 'Entertainment;Brain Games': 3314285.71,
 'Entertainment;Creativity': 4000000.0,
 'Entertainment;Music & Video': 6413333.33,
 'Events': 253542.22,
 'Finance': 1387692.48,
 'Food & Drink': 1924897.74,
 'Health & Fitness': 4188821.99,
 'House & Home': 1331540.56,
 'Libraries & Demo': 638503.73,
 'Lifestyle': 1412998.34,
 'Lifestyle;Pretend Play': 10000000.0,
 'Card': 3815462.5,
 'Arcade': 22888365.49,
 'Puzzle': 8302861.91,
 'Racing': 15910645.68,
 'Sports': 4596842.62,
 'Casual': 19569221.6,
 'Simulation': 3

In [45]:
avg_users_per_genre(free_apple_data, 12, 6)

{'Productivity': 21028.41,
 'Weather': 52279.89,
 'Shopping': 26919.69,
 'Reference': 74942.11,
 'Finance': 31467.94,
 'Music': 57326.53,
 'Utilities': 18684.46,
 'Travel': 28243.8,
 'Social Networking': 71548.35,
 'Sports': 23008.9,
 'Health & Fitness': 23298.02,
 'Games': 22812.92,
 'Food & Drink': 33333.92,
 'News': 21248.02,
 'Book': 39758.5,
 'Photo & Video': 28441.54,
 'Entertainment': 14029.83,
 'Business': 7491.12,
 'Lifestyle': 16485.76,
 'Education': 7003.98,
 'Navigation': 86090.33,
 'Medical': 612.0,
 'Catalogs': 4004.0}